apps/builder/app/shared/style-object-model.ts

```diff
@@ -0,0 +1,155 @@
+import { properties } from "@webstudio-is/css-data";
+import { StyleValue } from "@webstudio-is/css-engine";
+import type { Instance, StyleDecl, StyleSource } from "@webstudio-is/sdk";
+
+/**
+ *
+ * use cases where computing is called
+ *
+ * 1. compiler
+ *
+ * put the whole tree into the model once
+ * lazily compute each point in the tree reusing cached cascaded values
+ *
+ * 2. builder style panel
+ *
+ * property is updated within some point in the tree
+ * invalidate only changed properties in model
+ * lazily recompute changed properties
+ *
+ * custom property or color is updated within some point in the tree
+ * invalidate dependent properties in model
+ * lazily recompute changed and dependent properties
+ *
+ * any selection is changed
+ * invalidate model
+ * lazily recompute all properties
+ *
+ * 3. CSS Preview
+ *
+ * all properties are collected from instance selector
+ * property are iterated with own model and computed
+ *
+ */
+
+type InstanceSelector = string[];
+
+/**
+ * style selector is a full address in a tree
+ * and style data to extract computed style
+ */
+export type StyleSelector = {
+  instanceSelector: InstanceSelector;
+};
+
+type Style = Map<StyleDecl["property"], StyleDecl>;
+
+/**
+ * model contains all data and cache of computed styles
+ * and manages reactive subscriptions
+ */
+export type StyleObjectModel = {
```

Comment:
nit: I think you could call it StylesModel

Reply:
model is not the same as class
we have data and operations on it
looks like model
and again association with cssom

Reply:
cssom has methods on the model